clang/lib/CodeGen/CGExprCXX.cpp

```diff
@@ -2349,10 +2386,16 @@ RValue CodeGenFunction::EmitMatchPattern(const MatchPattern *Pattern,
   case MatchPattern::MatchPatternClass::BindingPatternClass: {
     auto *BinPat = static_cast<const BindingPattern *>(Pattern);
     const BindingDecl *D = BinPat->getBinding();
+    assert(D && "expected binding declaration");
     const VarDecl *Var = D->getHoldingVar();
-    assert(Var && "expected binding source");
 
-    EmitVarDecl(*Var);
+    if (Var)
+      EmitVarDecl(*Var);
+    else {
+      llvm_unreachable("Unsure what to do here");
+      // (void)EmitLValue(D->getBinding());
```

Comment:
Not sure if there's anything to be done here. This would be the case where there is no holding variable underneath the binding. For example:
```cpp
void f(std::tuple<int, double> t) {
  auto [x, y] = t;
}
```
`x` is a binding that has `getHoldingVar()`: `std::tuple_element_t<0, std::tuple<int, double>>& __v = std::get<0>(t);` and `getBinding()` which is a `DeclRefExpr` to `__v`, whereas
```cpp
void f(const int (&is)[2]) {
  auto [x, y] = is;
}
```
Here, `x` is a binding where `getHoldingVar()` is null, and `getBinding()` returns the expression `is[0]`.

Not sure if that's context/information that you needed or not, and I don't know what the answer is as to what needs to be done here, but there you go 😄

Reply:
On my current investigating example `[let x, 0]` this is a temporary assert to hold the current investigation, when I emit the decomposeddecl, it works if I skip this (because no DeclRefExpr will show up), but then I get a similar problem while matching to materialize the binding on `== 0`, so there's more to it that needs investigating.